### ISSUE TITLE
Near Future Config Issue- Reactor

### ISSUE THREAD

**Issue by ironhelm** [Kerbalism/Kerbalism#461]
Noticed an issue that I've found is caused by Kerbalism's config for Near Future Electrical. May be related to issue #305. Kerbalism removes NFE's reactors ability to generate; both kerbalism and Near Future's own electrical panel show them as outputting no power. This is fixed by removing the config for Near Future, but this also removes the other adaptations for Near Future like radiation from reactors.

Have attached screenshot of only errors I could find related to issue.

<img width="459" alt="Screen Shot 2019-07-10 at 8 45 59 PM" src="https://user-images.githubusercontent.com/30600732/61016320-d5bca080-a354-11e9-8214-0169b27fa27b.png">

**Comment by UsernamesAreAlwaysTaken** [comment]
I believe the log spam is due to NFE's plugin.

We have to replace the module and clone the functionality with our own module, to allow us to add Kerbalism's features to the reactor, including background simulation (EC generation in the background).

I'll see if I can do anything about it, but no promises. Most likely this Issue should be posted in NFE's repo, as we can;t really do much about it.

**Comment by gotmachine** [comment]
The log spam is actually a Kerbalism bug (that is fixed in this commit : 497f6c2512fa9160c40e0a10cc669a72d3dd17a2). Anyway I doubt It's what causing the NFE issues.

The latest versions of Near Future Electrical / Kerbal Atomics (and maybe others NF mods) depend on DynamicBatteryStorage 2, which in the current state of things **will** cause issues with Kerbalism.
I suggest removing it from you install and see 

I have some far plans to redo support of the NF / Nerteas mods trough our own partmodules (basically a process with internal heat management capabilities), which is the only sustainable way to go as supporting other mods modules is a lot of work and a never-ending task.

Until then, maybe we can find workarounds.

**Comment by ChrisAdderley** [comment]
> The latest versions of Near Future Electrical / Kerbal Atomics (and maybe others NF mods) depend on DynamicBatteryStorage 2, which in the current state of things will cause issues with Kerbalism.

Can you elaborate on this? If we can avoid stepping on each others' toes that would be nice - I don't want to make DBS work with Kerbalism, but it would be ideal if neither caused each other to break. 

> I have some far plans to redo support of the NF / Nerteas mods trough our own partmodules (basically a process with internal heat management capabilities), which is the only sustainable way to go as supporting other mods modules is a lot of work and a never-ending task.

It seems generally where Kerbalism runs into problems with mods is electricity management and your need to to run processes to consume/generate electricity in the background. It sounds to me that instead of replacing processes with those you own you might consider writing adapter modules that other mods can hook into to deliver appropriate information to your background simulators.

For example, I think it would greatly reduce your maintenance burden if you provided a way for me to tell Kerbalism that `this part is producing 450 EC/s`. You can then handle all the appropriate background processing and the other mod author handles all the rest of their module however they want as long as they ensure that the value being delivered is correct. The interaction surface is greatly minimized in this way. 

Obviously a simplification, but I get a lot of "support kerbalism plz" requests, which I naturally have to decline as the Kerbalism approach doesn't involve me at all - it's fairly non-cooperative.

**Comment by gotmachine** [comment]
>The latest versions of Near Future Electrical / Kerbal Atomics (and maybe others NF mods) depend on DynamicBatteryStorage 2, which in the current state of things will cause issues with Kerbalism.

Well I've not tested it, it's a supposition based on a very quick look at the DynamicBatteryStorage  source. Anyway we already replace all the partmodules from NFE with our own, but I think the simple presence of the DynamicBatteryStorage plugin can causes issues. In any case, the UI part of DynamicBatteryStorage  is 100% redundant with our own planner and even if ot doesn't break, it won't be accurate when Kerbalism is installed.

To solve this in a brutal yet efficient way, I'm considering simply unloading the assembly from the game when it's detected (with a prompt to the user). This "plugin blacklist" feature is something that we need anyway, because, as you correctly stated, there are many plugins that are incompatible with Kerbalism, and relying on players to know which ones is just hopeless.

>It seems generally where Kerbalism runs into problems with mods is electricity management and your need to to run processes to consume/generate electricity in the background.

Actually the problem is with all resources, and also with loaded vessels. We completely replace the stock resource handling and use the same simulation, no matter if the vessel is loaded or not.

> It sounds to me that instead of replacing processes with those you own you might consider writing adapter modules that other mods can hook into to deliver appropriate information to your background simulators.

>For example, I think it would greatly reduce your maintenance burden if you provided a way for me to tell Kerbalism that this part is producing 450 EC/s.

We are considering it. However, it's not that simple. Doing background simulation imply that anything that happen on the partmodule can be replicated without it, in a static class, with only the persisted information available (and eventually a reference to the part prefab). An example of something that is not available in this context is the stock core heat system.

Also, we have some projects for our own thermal control system. Having support code (from you, or from us) for the resource management + the thermal aspects of your partmodules will be more work than just doing our own implementation.

> Obviously a simplification, but I get a lot of "support kerbalism plz" requests, which I naturally have to decline as the Kerbalism approach doesn't involve me at all - it's fairly non-cooperative.

It's somewhat of a design decision at the core of the mod, to have our own unified way of doing things. But it's indeed not ideal when it comes to supporting mods that have their own custom things going on.

All this being said, in my experience, it's less problematic in the long run to override completely other mods modules with our own. No matter if Kerbalism has custom code for supporting other mods code or the other way around, it will break a lot more often when either mod updates.

With our own partmodules, it's usually just a matter of keeping the configs up to date with new/changed parts, and that can sometimes be automated fairly well trough MM patches.

All this being said, your parts are fantastic, I'm a huge fan.

**Comment by ChrisAdderley** [comment]
Yes, that all makes complete sense from your perspective. As long as you accept the responsibility of shunting all Kerbalism-related support questions to you I have no issues. It may result in some player unhappiness who want to use the systems together but that is your call. 

A couple notes:

> Well I've not tested it, it's a supposition based on a very quick look at the DynamicBatteryStorage source. Anyway we already replace all the partmodules from NFE with our own, but I think the simple presence of the DynamicBatteryStorage plugin can causes issues. In any case, the UI part of DynamicBatteryStorage is 100% redundant with our own planner and even if ot doesn't break, it won't be accurate when Kerbalism is installed.
> 
> To solve this in a brutal yet efficient way, I'm considering simply unloading the assembly from the game when it's detected (with a prompt to the user). This "plugin blacklist" feature is something that we need anyway, because, as you correctly stated, there are many plugins that are incompatible with Kerbalism, and relying on players to know which ones is just hopeless.

If you can just blacklist this, it makes both our problems go away, so that is ideal.

> We are considering it, there as been a similar request from @sswelm . However, it's not that simple. Doing background simulation imply that anything that happen on the partmodule can be replicated without it, in a static class, with only the persisted information available (and eventually a reference to the part prefab). An example of something that is not available in this context is the stock core heat system.

I think in this case you break things into first-party support (things that Kerbalism has ingested and has full control over the background simulation of) and second party (everything else). You are making an assumption that these modules are a black box, which handle their own internals such that the value is kept updated. As long as they conform to the basic interface things should be ok. 

This is indeed more work than just blacklisting or replacing, but it does allow a bit more player flexibility.

**Comment by Shiolle** [comment]
What do you think about separating background simulation logic and heat mechanics (if it is ever written) into separate mods which could both Kerbalism and other mods could use? Similar to, for example, modular flight integrator.

You would still control the way things happen, but you no longer need to support their configs and change compatibility configs when they introduce new features. On the other hand, there won't be several different background simulation implementations all conflicting with each other, making huge problems for players who want to use mods depending on these implementations.

**Comment by gotmachine** [comment]
> What do you think about separating [...] mechanics [...] into separate mods which both Kerbalism and other mods could use?

While splitting some features of Kerbalism in "optional sub-mods" is something that we are considering, those sub-mods won't be standalone, they will depend on the core Kerbalism. And even that is huge a mount of work, and will probably never happen.

> you no longer need to support their configs and change compatibility configs when they introduce new features.

There is nothing currently preventing mods to make support configs for Kerbalism, using the Kerbalism modules and framework instead of the stock ones (or their own). Some of them do.

> there won't be several different background simulation implementations all conflicting with each other

Unfortunately, in my experience, plugin authors go their own way and don't like to be dependent on other mods. Which is 100% rational. Especially for background simulation, which is complex and very likely to cause issues. There has been previous attempts at providing a background simulation framework, they all failed for this reason.

**Comment by gotmachine** [comment]
I'm implementing a basic radiator partmodule to replicate the dumb core heat stock system, nothing fancy.
Using that, I will try to come up with a basic heat-dependent resource converter to use with the NFE reactors, something that is at least somewhat balanced instead of the awful thing we have currently.

**Comment by Aebestach** [comment]
Issues related to SystemHeat and its associated mods should have been resolved in Kerbalism 3.40. Some issues are relatively outdated, so any new related problems can be addressed by reopening a new issue. However, if Kerbalism 3.40 (specifically the SystemHeat portion) is proven unreliable, then this issue will be re-enabled.
